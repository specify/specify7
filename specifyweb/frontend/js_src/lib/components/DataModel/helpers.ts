import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
