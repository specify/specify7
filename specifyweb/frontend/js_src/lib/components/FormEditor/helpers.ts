import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
