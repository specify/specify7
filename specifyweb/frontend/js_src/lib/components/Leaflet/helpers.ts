/**
 * Utility functions for getting locality data from locality resource and
 * workbench dataset
 *
 * @module
 */

import { Coord } from '../../utils/latLong';
import type { IR, RA } from '../../utils/types';
import type { MappingPath } from '../WbPlanView/Mapper';
import {
  mappingPathToString,
  valueIsTreeRank,
} from '../WbPlanView/mappingHelpers';
import { mappingLocalityColumns, requiredLocalityColumns } from './config';

export type Field<T extends Readonly<unknown>> = {
  readonly headerName: string;
  readonly value: T;
};

export type LocalityData = IR<Field<number | string>>;

export const getLocalityField = (
  row: RA<string>,
  headers: RA<string>,
  localityColumns: IR<string>,
  fieldName: string
): string => row[headers.indexOf(localityColumns[fieldName] ?? -1)] ?? '';

export function formatCoordinate(coordinate: string): number {
  if (coordinate === '' || coordinate === '0') return 0;

  const parsedCoordinate = Coord.parse(coordinate)?.toDegs();
  return parsedCoordinate === undefined
    ? 0
    : parsedCoordinate.components[0] * parsedCoordinate.sign;
}

export const findRanksInMappings = (
  mappingPaths: RA<MappingPath>
): RA<{ readonly groupName: string; readonly treeRankLocation: number }> =>
  mappingPaths
    .map((mappingPath) => ({
      mappingPath,
      treeRankLocation: mappingPath.findIndex(valueIsTreeRank),
    }))
    .map(({ mappingPath, treeRankLocation }) =>
      treeRankLocation === -1
        ? { groupName: '', treeRankLocation }
        : {
            treeRankLocation,
            groupName: mappingPathToString(
              mappingPath.slice(0, treeRankLocation)
            ),
          }
    );

export const getLocalityData = (
  localityColumns: IR<string>,
  getField: (fieldName: string) => Field<string>,
  formatCoordinate: (fieldName: string) => Field<number>
): LocalityData | false =>
  requiredLocalityColumns.every((fieldName) => getField(fieldName).value !== '')
    ? {
        ...Object.fromEntries(
          Object.keys(localityColumns)
            .filter(
              (columnName) => !mappingLocalityColumns.includes(columnName)
            )
            .map((columnName) => [columnName, getField(columnName)])
        ),
        'locality.latitude1': formatCoordinate('locality.latitude1'),
        'locality.longitude1': formatCoordinate('locality.longitude1'),
        ...(getField('locality.latitude2').value !== '' &&
        getField('locality.longitude2').value !== '' &&
        getField('locality.latlongtype').value.toLowerCase() !== 'point'
          ? {
              'locality.latitude2': formatCoordinate('locality.latitude2'),
              'locality.longitude2': formatCoordinate('locality.longitude2'),
              'locality.latlongtype': {
                value: ['line', ''].includes(
                  getField('locality.latlongtype').value.toLowerCase()
                )
                  ? 'line'
                  : 'rectangle',
                headerName: getField('locality.latlongtype').headerName,
              },
            }
          : {}),
        'locality.latlongaccuracy': isValidAccuracy(
          getField('locality.latlongaccuracy').value
        )
          ? getField('locality.latlongaccuracy')
          : { value: '', headerName: '' },
      }
    : false;

export function isValidAccuracy(
  latlongaccuracy: string | undefined
): latlongaccuracy is string {
  try {
    return (
      latlongaccuracy !== undefined &&
      !Number.isNaN(Number.parseFloat(latlongaccuracy)) &&
      Number.parseFloat(latlongaccuracy) >= 1
    );
  } catch {
    return false;
  }
}
