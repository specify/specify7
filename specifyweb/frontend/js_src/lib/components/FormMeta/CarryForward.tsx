import React from 'react';

import { useBooleanState } from '../../hooks/useBooleanState';
import { useId } from '../../hooks/useId';
import { commonText } from '../../localization/common';
import { formsText } from '../../localization/forms';
import { interactionsText } from '../../localization/interactions';
import { schemaText } from '../../localization/schema';
import { wbPlanText } from '../../localization/wbPlan';
import { f } from '../../utils/functools';
import type { IR, RA } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { sortFunction, toggleItem } from '../../utils/utils';
import { H3, Ul } from '../Atoms';
import { Button } from '../Atoms/Button';
import { Form, Input, Label } from '../Atoms/Form';
import { icons } from '../Atoms/Icons';
import { Submit } from '../Atoms/Submit';
import { getFieldsToClone, getUniqueFields } from '../DataModel/resource';
import type { LiteralField, Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables } from '../DataModel/tables';
import { NO_CLONE } from '../Forms/ResourceView';
import { Dialog } from '../Molecules/Dialog';
import { userPreferences } from '../Preferences/userPreferences';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';

/**
 * Fields to always carry forward (unless "Deselect All" is pressed), but not
 * show in the UI.
 */
const invisibleCarry = new Set([
  'collection',
  'collections',
  'definition',
  'definitionItem',
  'discipline',
  'disciplines',
  'division',
  'divisions',
  // There is literally a database field with a typo in its name (in Address table)
  'insitutions',
  'institution',
  'institutions',
  'scope',
]);

/** Search for all dependent fields using a suffix */
const dependentFieldSeeker = (suffix: string): IR<string> =>
  Object.fromEntries(
    Object.values(genericTables)
      .flatMap(({ literalFields }) =>
        literalFields.filter((v) => v.name.toLowerCase().endsWith(suffix))
      )
      .map(
        (v) =>
          [
            v.name,
            v.table.getField(v.name.slice(0, -suffix.length))?.name,
          ] as const
      )
      .filter(([_dependent, source]) => typeof source === 'string')
  ) as IR<string>;

export const strictDependentFields = f.store<IR<string>>(() => ({
  // Fields like endDatePrecision
  ...dependentFieldSeeker('precision'),
  // Fields like endDateVerbatim
  ...dependentFieldSeeker('verbatim'),
  // Fields like endDepthUnit
  ...dependentFieldSeeker('unit'),
}));

/**
 * Dependent -> Source
 * When value in one field is based on another, don't show the dependent field in
 * the UI, but copy its carry over settings from the source field
 */
export const dependentFields = f.store<IR<string>>(() => ({
  lat1text: 'latitude1',
  lat2text: 'latitude2',
  long1text: 'longitude1',
  long2text: 'longitude2',
  ...strictDependentFields(),
}));

export function CarryForwardConfig({
  table,
  parentTable,
  type,
}: {
  readonly table: SpecifyTable;
  readonly parentTable: SpecifyTable | undefined;
  readonly type: 'button' | 'cog';
}): JSX.Element | null {
  const [isOpen, handleOpen, handleClose] = useBooleanState();
  const [globalEnabled, setGlobalEnabled] = userPreferences.use(
    'form',
    'preferences',
    'enableCarryForward'
  );
  const isEnabled = globalEnabled.includes(table.name);
  const canChange = !NO_CLONE.has(table.name);

  return canChange ? (
    <>
      {type === 'button' ? (
        <Label.Inline className="rounded bg-[color:var(--foreground)]">
          <Input.Checkbox
            checked={isEnabled}
            onChange={(): void =>
              setGlobalEnabled(toggleItem(globalEnabled, table.name))
            }
          />
          {formsText.carryForwardEnabled()}
          <Button.Small
            className="ml-2"
            title={formsText.carryForwardSettingsDescription()}
            onClick={handleOpen}
          >
            {icons.cog}
          </Button.Small>
        </Label.Inline>
      ) : (
        <Button.Icon
          icon="cog"
          title={formsText.carryForwardSettingsDescription()}
          onClick={handleOpen}
        />
      )}
      {isOpen && (
        <CarryForwardConfigDialog
          parentTable={parentTable}
          table={table}
          onClose={handleClose}
        />
      )}
    </>
  ) : null;
}

const normalize = (fields: RA<string>): RA<string> =>
  Array.from(fields).sort(sortFunction(f.id));

function CarryForwardConfigDialog({
  table,
  parentTable,
  onClose: handleClose,
}: {
  readonly table: SpecifyTable;
  readonly parentTable: SpecifyTable | undefined;
  readonly onClose: () => void;
}): JSX.Element {
  const [showHiddenFields, setShowHiddenFields] = userPreferences.use(
    'form',
    'preferences',
    'carryForwardShowHidden'
  );

  const [globalConfig, setGlobalConfig] = userPreferences.use(
    'form',
    'preferences',
    'carryForward'
  );

  const uniqueFields = getUniqueFields(table);
  const defaultConfig = getFieldsToClone(table).filter(
    (fieldName) => !uniqueFields.includes(fieldName)
  );
  const isDefaultConfig = (fields: RA<string>): boolean =>
    JSON.stringify(normalize(fields)) ===
    JSON.stringify(normalize(defaultConfig));

  const config =
    (globalConfig[table.name] as RA<string> | undefined)?.filter(
      (fieldName) => !uniqueFields.includes(fieldName)
    ) ?? defaultConfig;

  const handleChange = (fields: RA<string>): void =>
    setGlobalConfig({
      ...globalConfig,
      [table.name]: isDefaultConfig(fields) ? undefined : fields,
    });

  const reverseRelationships = React.useMemo(
    () =>
      filterArray(
        parentTable?.relationships
          .filter(({ relatedTable }) => relatedTable === table)
          .flatMap(({ otherSideName }) => otherSideName) ?? []
      ),
    [parentTable, table]
  );

  const literalFields = table.literalFields.filter(
    ({ name, overrides, isVirtual }) =>
      !isVirtual &&
      (!overrides.isHidden || showHiddenFields) &&
      !invisibleCarry.has(name)
  );
  const relationships = table.relationships.filter(
    (field) =>
      !reverseRelationships.includes(field.name) &&
      !field.isVirtual &&
      (!field.overrides.isHidden || showHiddenFields) &&
      (field.isDependent() || !relationshipIsToMany(field)) &&
      !invisibleCarry.has(field.name)
  );

  const id = useId('form-carry-forward');
  return (
    <Dialog
      buttons={
        <>
          <Button.Success
            disabled={isDefaultConfig(config)}
            onClick={(): void =>
              handleChange(
                showHiddenFields
                  ? defaultConfig
                  : table.fields
                      .filter(
                        ({ name, isVirtual, overrides }) =>
                          !isVirtual &&
                          !reverseRelationships.includes(name) &&
                          (!overrides.isHidden || config.includes(name))
                      )
                      .map(({ name }) => name)
              )
            }
          >
            {interactionsText.selectAll()}
          </Button.Success>
          <Button.Success
            disabled={config.length === 0}
            onClick={(): void =>
              handleChange(
                // Don't deselect hidden fields if they are not visible
                showHiddenFields
                  ? []
                  : table.fields
                      .filter(
                        ({ name, isVirtual, overrides }) =>
                          !isVirtual &&
                          !reverseRelationships.includes(name) &&
                          overrides.isHidden &&
                          config.includes(name)
                      )
                      .map(({ name }) => name)
              )
            }
          >
            {interactionsText.deselectAll()}
          </Button.Success>
          <Submit.Info form={id('form')} onClick={handleClose}>
            {commonText.close()}
          </Submit.Info>
        </>
      }
      header={formsText.carryForwardTableSettingsDescription({
        tableName: table.label,
      })}
      onClose={handleClose}
    >
      <Form className="overflow-hidden" id={id('form')} onSubmit={handleClose}>
        <div className="flex flex-1 flex-col gap-2 overflow-y-auto">
          <CarryForwardCategory
            carryForward={config}
            fields={literalFields}
            header={schemaText.fields()}
            table={table}
            uniqueFields={uniqueFields}
            onChange={handleChange}
          />
          <CarryForwardCategory
            carryForward={config}
            fields={relationships}
            header={schemaText.relationships()}
            table={table}
            uniqueFields={uniqueFields}
            onChange={handleChange}
          />
        </div>
        <Label.Inline className="border-t pt-2 dark:border-neutral-700">
          <Input.Checkbox
            checked={showHiddenFields}
            onValueChange={setShowHiddenFields}
          />
          {wbPlanText.revealHiddenFormFields()}
        </Label.Inline>
      </Form>
    </Dialog>
  );
}

function CarryForwardCategory({
  header,
  table,
  fields,
  uniqueFields,
  carryForward,
  onChange: handleChange,
}: {
  readonly header: string;
  readonly table: SpecifyTable;
  readonly fields: RA<LiteralField | Relationship>;
  readonly uniqueFields: RA<string>;
  readonly carryForward: RA<string>;
  readonly onChange: (carryForward: RA<string>) => void;
}): JSX.Element | null {
  return fields.length > 0 ? (
    <>
      <H3>{header}</H3>
      <Ul>
        {fields.map((field) => {
          const isUnique = uniqueFields.includes(field.name);
          return (
            <li className="flex gap-1" key={field.name}>
              <Label.Inline
                title={
                  isUnique
                    ? formsText.carryForwardUniqueField()
                    : field.getLocalizedDesc()
                }
              >
                <Input.Checkbox
                  checked={f.includes(carryForward, field.name)}
                  disabled={isUnique}
                  onValueChange={(isChecked): void => {
                    const dependents = filterArray(
                      Object.entries(dependentFields())
                        .filter(([_dependent, source]) => source === field.name)
                        .map(([dependent]) => table.getField(dependent)?.name)
                    );
                    handleChange(
                      isChecked
                        ? f.unique([...carryForward, field.name, ...dependents])
                        : carryForward.filter(
                            (name) =>
                              name !== field.name && !dependents.includes(name)
                          )
                    );
                  }}
                />
                {field.label}
              </Label.Inline>
              {field.isRelationship && field.isDependent() && !isUnique ? (
                <CarryForwardConfig
                  parentTable={field.table}
                  table={field.relatedTable}
                  type="cog"
                />
              ) : undefined}
            </li>
          );
        })}
      </Ul>
    </>
  ) : null;
}
