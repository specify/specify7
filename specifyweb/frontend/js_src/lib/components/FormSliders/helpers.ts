import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
