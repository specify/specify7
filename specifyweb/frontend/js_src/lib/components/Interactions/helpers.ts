import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA, RestrictedTuple } from '../../utils/types';
import type { AnyInteractionPreparation } from '../DataModel/helperTypes';
import type { Tables } from '../DataModel/types';

export const interactionPrepTables: RestrictedTuple<
  AnyInteractionPreparation['tableName']
> = [
  'DisposalPreparation',
  'ExchangeInPrep',
  'ExchangeOutPrep',
  'GiftPreparation',
  'LoanPreparation',
];

type ExtractInteraction<T extends string> =
  T extends `${infer Prefix}Prep${string}` ? Prefix : never;

export type InteractionWithPreps = Tables[ExtractInteraction<
  AnyInteractionPreparation['tableName']
>];

export const interactionsWithPrepTables: RestrictedTuple<
  InteractionWithPreps['tableName']
> = ['Disposal', 'ExchangeIn', 'ExchangeOut', 'Gift', 'Loan'];

export type PreparationData = {
  readonly catalogNumber: string;
  readonly collectionObjectId: number;
  readonly taxon: string;
  readonly taxonId: number;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
};

export type PreparationRow = readonly [
  catalogNumber: string,
  collectionObjectId: number,
  taxonFullName: string,
  taxonId: number,
  preparationId: number,
  prepType: string,
  preparationCountAmt: number,
  amountLoaned: string | null,
  amountedGifted: string | null,
  amountExchanged: string | null,
  amountAvailable: string
];

export const getPrepsAvailableForLoanRs = async (
  recordSetId: number,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ isLoan }),
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
      isLoan,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
