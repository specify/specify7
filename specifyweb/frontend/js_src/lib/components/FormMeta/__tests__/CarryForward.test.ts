import { requireContext } from '../../../tests/helpers';
import { dependentFields, strictDependentFields } from '../CarryForward';

requireContext();

test('strictDependentFields', () =>
  expect(strictDependentFields()).toMatchSnapshot());

test('dependentFields', () => expect(dependentFields()).toMatchSnapshot());
