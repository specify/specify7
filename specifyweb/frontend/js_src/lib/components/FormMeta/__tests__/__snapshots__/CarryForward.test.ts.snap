// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependentFields 1`] = `
{
  "GGBN_concentrationUnit": "GGBN_concentration",
  "GGBN_volumeUnit": "GGBN_volume",
  "GGBN_weightUnit": "GGBN_weight",
  "borrowDatePrecision": "borrowDate",
  "catalogedDatePrecision": "catalogedDate",
  "catalogedDateVerbatim": "catalogedDate",
  "completedDatePrecision": "completedDate",
  "curatorApprovalDatePrecision": "curatorApprovalDate",
  "date1Precision": "date1",
  "date2Precision": "date2",
  "date3Precision": "date3",
  "date4Precision": "date4",
  "date5Precision": "date5",
  "dateOfBirthPrecision": "dateOfBirth",
  "dateOfDeathPrecision": "dateOfDeath",
  "determinedDatePrecision": "determinedDate",
  "embargoReleaseDatePrecision": "embargoReleaseDate",
  "embargoStartDatePrecision": "embargoStartDate",
  "endDatePrecision": "endDate",
  "endDateVerbatim": "endDate",
  "endDepthUnit": "endDepth",
  "endDepthVerbatim": "endDepth",
  "examDatePrecision": "examDate",
  "extractionDatePrecision": "extractionDate",
  "inventoryDatePrecision": "inventoryDate",
  "lat1text": "latitude1",
  "lat2text": "latitude2",
  "long1text": "longitude1",
  "long2text": "longitude2",
  "maxUncertaintyEstUnit": "maxUncertaintyEst",
  "preparedDatePrecision": "preparedDate",
  "sequenceDatePrecision": "sequenceDate",
  "startDatePrecision": "startDate",
  "startDateVerbatim": "startDate",
  "startDepthUnit": "startDepth",
  "startDepthVerbatim": "startDepth",
  "treatmentCompDatePrecision": "treatmentCompDate",
}
`;

exports[`strictDependentFields 1`] = `
{
  "GGBN_concentrationUnit": "GGBN_concentration",
  "GGBN_volumeUnit": "GGBN_volume",
  "GGBN_weightUnit": "GGBN_weight",
  "borrowDatePrecision": "borrowDate",
  "catalogedDatePrecision": "catalogedDate",
  "catalogedDateVerbatim": "catalogedDate",
  "completedDatePrecision": "completedDate",
  "curatorApprovalDatePrecision": "curatorApprovalDate",
  "date1Precision": "date1",
  "date2Precision": "date2",
  "date3Precision": "date3",
  "date4Precision": "date4",
  "date5Precision": "date5",
  "dateOfBirthPrecision": "dateOfBirth",
  "dateOfDeathPrecision": "dateOfDeath",
  "determinedDatePrecision": "determinedDate",
  "embargoReleaseDatePrecision": "embargoReleaseDate",
  "embargoStartDatePrecision": "embargoStartDate",
  "endDatePrecision": "endDate",
  "endDateVerbatim": "endDate",
  "endDepthUnit": "endDepth",
  "endDepthVerbatim": "endDepth",
  "examDatePrecision": "examDate",
  "extractionDatePrecision": "extractionDate",
  "inventoryDatePrecision": "inventoryDate",
  "maxUncertaintyEstUnit": "maxUncertaintyEst",
  "preparedDatePrecision": "preparedDate",
  "sequenceDatePrecision": "sequenceDate",
  "startDatePrecision": "startDate",
  "startDateVerbatim": "startDate",
  "startDepthUnit": "startDepth",
  "startDepthVerbatim": "startDepth",
  "treatmentCompDatePrecision": "treatmentCompDate",
}
`;
